### PR TITLE
Changing pandora index format

### PR DESCRIPTION
**Pull request by leoisl**
This PR changes the pandora index from a set of files in a directory structure to a single, compressible and indexable `zip`
file (`pandora` indexes now have the suffix `.panidx.zip`). This is now the single file that is produced by the
`pandora index` command and is required as argument to all the other `pandora` commands. This index is self contained in 
the sense that it encodes all the information and metadata about it (e.g. which PRGs were used to create it, window and 
kmer size, etc). This new index provide the infrastructure for the next features and simplifies working with large 
reference pangenome collections, with a few million PRGs. These changes will be released as `pandora v0.11.0`.

Closes https://github.com/rmcolq/pandora/issues/308 https://github.com/rmcolq/pandora/issues/307 https://github.com/rmcolq/pandora/issues/306

## Breakdown of main changes
Sorry that this is another big PR, but half of the changes can be ignored as they are just updating the example data. Here is a breakdown of the main changes:
  * All the 29 files in the `example/` dir can be ignored, I just updated the example files to the latest version;
  * C++ version upgraded from 11 to 14 (`CMakeLists.txt`);
  * All `*_main(.h/.cpp)` files were changed to receive a pandora index (`*.panidx.zip`) instead of a fasta file containing PRGs;
  * `compare_main(.h/.cpp)`, `map_main(.h/.cpp)`, `discover_main(.h/.cpp)`, `seq2path_main(.h/.cpp)` all changed in a similar way to remove CLI parameters `-w` and `-k` (we now get this metadata from the index);
  * We have several refactorings to the `pandora` index implementation (`index(.h/.cpp)`);
  * We added a memory-efficient way to load PRGs when indexing, where we don't need to load all PRGs at once to index them, but just load on demand (`localPRG_reader(.h/.cpp)`);
  * Removed `merge_index` subcommand (`merge_index_main(.h/.cpp)`);
  * Added classes to deal with Zip files (`zip_file(.h/.cpp)`);

### Changed

- The `pandora` index changed from a set of files in a directory structure to a single, compressible and indexable `zip`
file (`pandora` indexes now have the suffix `.panidx.zip`). This is now the single file that is produced by the
`pandora index` command and is required as argument to all the other `pandora` commands. This index is self contained in 
the sense that it encodes all the information and metadata about it (e.g. which PRGs were used to create it, window and 
kmer size, etc). This new index provide the infrastructure for the next features and simplifies working with large 
reference pangenome collections, with a few million PRGs. This new index breaks backwards compatibility with previous 
`pandora` versions. The structure of this zip archive is as follows:
  * `_prgs`: The PRGs themselves used as input to create this index;
  * `_prg_names`: The names of the PRGs;
  * `_prg_min_path_lengths`: the length of the shortest path through each PRG;
  * `_minhash`: the minimizer hash data structure;
  * `_metadata`: metadata about the index (first line is window size, second is kmer size);
  * `*.gfa`: the several GFA files describing the minimizing kmer graph for each PRG;
- Minimum C++ standard upgraded from `C++11` to `C++14`;

### Removed
- Removed CLI parameters `-w` and `-k` from the following `pandora` subcommands: `compare`, `discover`, `map`,
`seq2path`;   
- Removed `merge_index` subcommand;

### Fixed
- Several refactoring to the `pandora` index implementation;

### Added
- A memory-efficient way to load PRGs when indexing, where we don't need to load all PRGs at once to index them, but 
just load on demand;

### CODE REVIEW

File: include/compare_main.h [hunk 1]
```diff
@@ -29,11 +29,9 @@ using std::set;
 using std::vector;
 
 struct CompareOptions {
-    fs::path prgfile;
+    fs::path pandora_index_file;
```
Comment (mbhall88) [inline include/compare_main.h:32]: Maybe just simpler as `index_file`? It's kind of obvious it's a pandora index no? Anyway, that's a very minor thing. Don't worry about changing it if you can't be bothered

File: CMakeLists.txt [hunk 2]
```diff
@@ -12,20 +12,17 @@ HunterGate(
 
 # project configuration
 set(PROJECT_NAME_STR pandora)
-project(${PROJECT_NAME_STR} VERSION "0.10.0" LANGUAGES C CXX)
+project(${PROJECT_NAME_STR} VERSION "0.11.0" LANGUAGES C CXX)
```
Comment (mbhall88) [inline CMakeLists.txt:15]: We're okay to stay on 0.10.0 as we are still in alpha release, which basically says we are in the process of adding features for 0.10.0. Best not to skip 0.10.0 as we never did a stable release yet.

**Review by mbhall88 — CHANGES_REQUESTED** [review]
Can't say I checked every bit of code, but love the changes. Especially getting rid of the need for k and w in map and compare and discover!

The one change I would recommend is relating to the version. We should stay on 0.10.0 but bump to 0.10.0-alpha.1 (see my comment in CMakeList.txt for more details)

File: CMakeLists.txt (hunk 2 above)
Comment (leoisl) [reply CMakeLists.txt]: Done in https://github.com/rmcolq/pandora/pull/318/commits/a6f7db0ef50a3926d3c3e8ae971cb10941b103c7

File: include/compare_main.h (hunk 1 above)
Comment (leoisl) [reply include/compare_main.h]: Done in https://github.com/rmcolq/pandora/pull/318/commits/2631de2697d65def8622931117686db46f6b213b

**Comment by leoisl** [comment]
Thanks for the comments, both were addressed! However, I am hesitant to pre-release this, as I've not tested on real data, and we're delaying writing unit tests for the new code for later. The main contribution of this PR is to provide infrastructure to the lazy loading of PRGs feature, which is essential to roundhound. This feature should be finished somewhere next week, and then I will test it with roundhound dataset, and will be more confident on pre-releasing it.

File: CMakeLists.txt [hunk 3]
```diff
@@ -12,7 +12,7 @@ HunterGate(
 
 # project configuration
 set(PROJECT_NAME_STR pandora)
-project(${PROJECT_NAME_STR} VERSION "0.11.0" LANGUAGES C CXX)
+project(${PROJECT_NAME_STR} VERSION "0.10.0" LANGUAGES C CXX)
```
Comment (mbhall88) [inline CMakeLists.txt:15]: Just wondering, is it possible to make the version 0.10.0-alpha.1 here? I tried looking at the CMake docs and it wasn't super clear...

**Review by mbhall88 — APPROVED** [review]

**Comment by mbhall88** [comment]
> However, I am hesitant to pre-release this, as I've not tested on real data, and we're delaying writing unit tests for the new code for later.

Sure. Do a pre-release whenever you think it makes sense (i.e. when there is a complete set/functionaly for the new features)

File: CMakeLists.txt (hunk 3 above)
Comment (leoisl) [reply CMakeLists.txt:15]: I've tried this before and was unable to do it, but will retry just in case

Comment (leoisl) [reply CMakeLists.txt:15]: I didn't manage to do it, there is a 5-year open issue in CMake git repo asking for this feature, but hasn't been finished yet: https://gitlab.kitware.com/cmake/cmake/-/issues/16716 . So I am keeping the version as is I guess...

Comment (leoisl) [reply CMakeLists.txt:15]: Oh, we can actually use an extra number for the tweak version, so won't be able to set `0.10.0-alpha.1` but can set `0.10.0.1`.

Comment (leoisl) [reply CMakeLists.txt]: Just did this in https://github.com/rmcolq/pandora/pull/318/commits/524c7ae755e1a711c717ec3a9bf2aa3c5292399f , hope is ok. I am then merging this PR. If there is a problem, please tell me in slack or raise an issue.

**Pull request merged**